packages/@lwc/template-compiler/src/codegen/index.ts

```diff
@@ -122,6 +127,15 @@ function transform(codeGen: CodeGen): t.Expression {
             let expr;
 
             if (isElement(child)) {
+                // When the same element have if and forEach/forOf directives, we must create two scopes.
+                if (child.if) {
+                    codeGen.createScope();
```

Comment:
I would rename most of the references from `scope` to `block`. While an iteration creates a new lexical scope with the item and the index, the if block doesn't. This is why I think the term `block` makes more sense in this context.